File: src/main/java/io/github/divyeshmangla/timetable/config/ConfigLoader.java
package io.github.divyeshmangla.timetable.config;

import io.github.divyeshmangla.timetable.model.Subject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.yaml.snakeyaml.Yaml;

import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

public final class ConfigLoader {
    private static final Logger LOGGER = LoggerFactory.getLogger(ConfigLoader.class);
    private static final String INIT_CONFIG_FLAG = "--init-config";

    private final String configFileName;

    public ConfigLoader(String configFileName) {
        this.configFileName = configFileName;
    }

    /**
     * Resolves config: local file preferred, fallback to bundled.
     * @return InputStream that must be closed by caller
     */
    public InputStream resolve() throws IOException {
        Path localConfig = Path.of(configFileName);

        if (Files.exists(localConfig)) {
            LOGGER.debug("Using local {}", configFileName);
            return Files.newInputStream(localConfig);
        }

        LOGGER.debug("Using bundled {}", configFileName);
        return getBundledConfig();
    }

    /**
     * Handles --init-config flag: copies bundled config to working directory.
     * @return true if flag was handled (caller should exit), false otherwise
     */
    public boolean handleInitFlag(String[] args) throws IOException {
        if (args == null || !Arrays.asList(args).contains(INIT_CONFIG_FLAG)) {
            return false;
        }

        Path localConfig = Path.of(configFileName);
        if (Files.exists(localConfig)) {
            LOGGER.info("{} already exists", configFileName);
            return true;
        }

        try (InputStream in = getBundledConfig()) {
            Files.copy(in, localConfig);
            LOGGER.info("{} created successfully", configFileName);
        }

        return true;
    }

    /**
     * Parses config from input stream.
     */
    public static Config parse(InputStream in) {
        Yaml yaml = new Yaml();
        Map<String, Object> root = yaml.load(in);

        @SuppressWarnings("unchecked")
        Map<String, Object> timetable = (Map<String, Object>) root.get("timetable");
        if (timetable == null) throw new IllegalStateException("Missing timetable section in config");


        Object urlObj = timetable.get("url");
        if (urlObj == null) throw new IllegalStateException("Missing timetable.url in config");

        String timetableUrl = urlObj.toString();

        @SuppressWarnings("unchecked")
        Map<String, Object> subjectsNode = (Map<String, Object>) root.get("subjects");
        List<Subject> subjects = parseSubjects(subjectsNode);

        return new Config(List.copyOf(subjects), timetableUrl);
    }

    private static List<Subject> parseSubjects(Map<String, Object> subjectsNode) {
        List<Subject> subjects = new ArrayList<>();

        if (subjectsNode != null) {
            for (var entry : subjectsNode.entrySet()) {
                String code = entry.getKey();

                @SuppressWarnings("unchecked")
                var data = (Map<String, Object>) entry.getValue();
                var name = (String) data.get("name");
                var abbr = (String) data.get("abbreviation");

                subjects.add(new Subject(code, name, abbr));
            }
        }

        return subjects;
    }

    private InputStream getBundledConfig() throws IOException {
        InputStream in = ConfigLoader.class
                .getClassLoader()
                .getResourceAsStream(configFileName);

        if (in == null) throw new IOException("Bundled " + configFileName + " not found in classpath");

        return in;
    }
}

File: internal/config/config_loader.go
package config

import (
	"errors"
	"fmt"
	"io"
	"os"
	"slices"

	"github.com/DivyeshMangla/tiet-timetable/internal/model"
	"github.com/DivyeshMangla/tiet-timetable/internal/types"
	"gopkg.in/yaml.v3"
)

const initConfigFlag = "--init-config"

var ErrMissingTimetableURL = errors.New("missing timetable.url in config")

type ConfigLoader struct {
	configFileName string
}

func NewConfigLoader(configFileName string) *ConfigLoader {
	return &ConfigLoader{configFileName: configFileName}
}

func (cl *ConfigLoader) Load() (*Config, error) {
	reader, err := cl.Resolve()
	if err != nil {
		return nil, err
	}
	defer reader.Close()
	return cl.Parse(reader)
}

func (cl *ConfigLoader) Resolve() (io.ReadCloser, error) {
	if cl.hasLocalConfig() {
		return cl.openLocalConfig()
	}
	return cl.getBundledConfig()
}

func (cl *ConfigLoader) HandleInitFlag(args []string) (bool, error) {
	if !slices.Contains(args, initConfigFlag) {
		return false, nil
	}
	if cl.hasLocalConfig() {
		fmt.Printf("%s already exists\n", cl.configFileName)
		return true, nil
	}
	return true, cl.copyBundledToLocal()
}

func (cl *ConfigLoader) Parse(in io.Reader) (*Config, error) {
	var root struct {
		Timetable struct {
			URL string `yaml:"url"`
		} `yaml:"timetable"`
		Subjects map[string]struct {
			Name         string `yaml:"name"`
			Abbreviation string `yaml:"abbreviation"`
		} `yaml:"subjects"`
	}
	if err := yaml.NewDecoder(in).Decode(&root); err != nil {
		return nil, fmt.Errorf("failed to decode YAML: %w", err)
	}
	if root.Timetable.URL == "" {
		return nil, ErrMissingTimetableURL
	}
	subjects := make([]model.Subject, 0, len(root.Subjects))
	for code, data := range root.Subjects {
		subjects = append(subjects, model.NewSubject(
			types.SubjectCode(code),
			types.SubjectName(data.Name),
			types.SubjectAbbr(data.Abbreviation),
		))
	}
	return &Config{
		Subjects:     subjects,
		TimetableURL: root.Timetable.URL,
	}, nil
}

func (cl *ConfigLoader) hasLocalConfig() bool {
	_, err := os.Stat(cl.configFileName)
	return err == nil
}

func (cl *ConfigLoader) openLocalConfig() (io.ReadCloser, error) {
	file, err := os.Open(cl.configFileName)
	if err != nil {
		return nil, fmt.Errorf("failed to open local config %q: %w", cl.configFileName, err)
	}
	return file, nil
}

func (cl *ConfigLoader) getBundledConfig() (io.ReadCloser, error) {
	file, err := FS.Open("config.yaml")
	if err != nil {
		return nil, fmt.Errorf("bundled config not found: %w", err)
	}
	return file, nil
}

func (cl *ConfigLoader) copyBundledToLocal() error {
	bundled, err := cl.getBundledConfig()
	if err != nil {
		return fmt.Errorf("failed to get bundled config: %w", err)
	}
	defer bundled.Close()

	local, err := os.Create(cl.configFileName)
	if err != nil {
		return fmt.Errorf("failed to create local config %q: %w", cl.configFileName, err)
	}
	defer local.Close()

	if _, err := io.Copy(local, bundled); err != nil {
		return fmt.Errorf("failed to copy config: %w", err)
	}
	fmt.Printf("%s created successfully\n", cl.configFileName)
	return nil
}
